tpm/ak.go

```diff
@@ -0,0 +1,204 @@
+package tpm
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"time"
+
+	"github.com/google/go-attestation/attest"
+
+	"go.step.sm/crypto/tpm/storage"
+)
+
+type AK struct {
+	Name      string
+	Data      []byte
+	CreatedAt time.Time
+
+	tpm *TPM
+}
+
+func (t *TPM) CreateAK(ctx context.Context, name string) (AK, error) {
+	result := AK{}
+	if err := t.Open(ctx); err != nil {
+		return result, fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer t.Close(ctx, false)
+
+	at, err := attest.OpenTPM(t.attestConfig)
+	if err != nil {
+		return result, fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer at.Close()
+
+	now := time.Now()
+
+	if name == "" {
+		// TODO: decouple the TPM key name from the name recorded in the storage? This might
+		// make it easier to work with the key names as a user; the TPM key name would be abstracted
+		// away. The key name in the storage can be different from the key stored with the key (which,
+		// to be far, isn't even used on Linux TPMs)
+		nameHex := make([]byte, 5)
+		if n, err := rand.Read(nameHex); err != nil || n != len(nameHex) {
+			return result, fmt.Errorf("rand.Read() failed with %d/%d bytes read and error: %w", n, len(nameHex), err)
+		}
+		name = fmt.Sprintf("%x", nameHex)
+	}
+
+	prefixedName := fmt.Sprintf("ak-%s", name)
+
+	akConfig := attest.AKConfig{
+		Name: prefixedName,
+	}
+	ak, err := at.NewAK(&akConfig)
+	if err != nil {
+		return result, err
+	}
+	defer ak.Close(at)
+
+	data, err := ak.Marshal()
+	if err != nil {
+		return result, err
+	}
+
+	storedAK := &storage.AK{
+		Name:      name,
+		Data:      data,
+		CreatedAt: now,
+	}
+
+	if err := t.store.AddAK(storedAK); err != nil {
+		return result, err
+	}
+
+	if err := t.store.Persist(); err != nil {
+		return result, err
+	}
+
+	return AK{Name: storedAK.Name, Data: storedAK.Data, CreatedAt: now, tpm: t}, nil
+}
+
+func (t *TPM) GetAK(ctx context.Context, name string) (AK, error) {
+	result := AK{}
+	if err := t.Open(ctx); err != nil {
+		return result, fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer t.Close(ctx, false)
+
+	ak, err := t.store.GetAK(name)
+	if err != nil {
+		return result, fmt.Errorf("error getting AK %q: %w", name, err)
+	}
+
+	return AK{Name: ak.Name, Data: ak.Data, CreatedAt: ak.CreatedAt, tpm: t}, nil
+}
+
+func (t *TPM) ListAKs(ctx context.Context) ([]AK, error) {
+	if err := t.Open(ctx); err != nil {
+		return nil, fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer t.Close(ctx, false)
```

Comment:
can close return an error? if so, you should return it.

I know it sounds counter-intuitive but this may very well be more important than the result you retrieve some times.

Reply:
Made the method private and am now catching the error on `close` in the defer. Errors originating from the main logic have priority, so I check if no error was set, and then override it with the one from `close` (if set).